Excluded or optional modules result in failed webpack build

Great webinar today, it was very informative! Thanks for doing it. I asked a question during it describing the following issue:

I currently exclude Buffer and crypto from my webpack build. I've seen this in various places being recommended to reduce build sizes. I do so with the config:

```javascript
node: {
  Buffer: false,
  crypto: false
}
```

We normally see warnings during our webpack build:

```
WARNING in ./webpack/scripts/third-party/pikaday.js
Module not found: Error: Cannot resolve module 'moment' in /Project/scripts/third-party
 @ ...

WARNING in ./seedrandom.js
Module not found: Error: Cannot resolve module 'crypto' in /Project/node_modules/@scope/seedrandom
@ ...
```

That first warning relates to the `moment` npm package which is an optional dependency of `pikaday`.

We can safely ignore these warnings, as we are confident they do not cause us any problems in production. However, when I pull in this plugin (with the default configs), the warnings turn into hard errors that fail the build.

```
ERROR in ./seedrandom.js
Module not found: Error: Cannot resolve module 'crypto' in /Project/node_modules/@scope/seedrandom
 @ ...

ERROR in ./webpack/scripts/third-party/pikaday.js
Module not found: Error: Cannot resolve module 'moment' in /Project/scripts/third-party
 @ ...
```

I've identified that if I alias these modules to `empty` modules for our development build, the errors go away, but I am worried about shipping such a change because it unnecessarily diverges our production and development builds.

Any advice here?

--- a/index.js
+++ b/index.js
@@ -166,6 +166,7 @@ function serializeDependencies(deps, parent) {
         recursive: dep.recursive,
         regExp: dep.regExp ? dep.regExp.source : null,
         async: dep.async,
+        optional: dep.optional,
         loc: flattenPrototype(dep.loc),
       };
     }

--- a/lib/deserialize-dependencies.js
+++ b/lib/deserialize-dependencies.js
@@ -23,6 +23,9 @@ function deserializeDependencies(deps, parent) {
       dep.critical = req.contextCritical;
       dep.async = req.async;
       dep.loc = req.loc;
+      if (req.optional) {
+        dep.optional = true;
+      }
       return dep;
     }
     if (req.constDependency) {
@@ -56,6 +59,9 @@ function deserializeDependencies(deps, parent) {
     }
     var dep = new HardModuleDependency(req.request);
     dep.loc = req.loc;
+    if (req.optional) {
+      dep.optional = true;
+    }
     return dep;
   }, this);
 }

--- a/tests/base-webpack-1.js
+++ b/tests/base-webpack-1.js
@@ -24,6 +24,8 @@ describe('basic webpack use - compiles identically', function() {
   itCompilesTwice('base-amd-context');
   itCompilesTwice('base-amd-code-split');
   itCompilesTwice('base-external');
+  itCompilesTwice('base-1dep-optional', {exportStats: true});
+  itCompilesTwice('base-context-optional', {exportStats: true});
 
 });
 

--- a/tests/fixtures/base-1dep-optional/index.js
+++ b/tests/fixtures/base-1dep-optional/index.js
@@ -1,0 +1,9 @@
+var fib;
+try {
+  fib = require('./fib');
+}
+catch (_) {
+  fib = function(v) {return v;};
+}
+
+console.log(fib(3));

--- a/tests/fixtures/base-1dep-optional/webpack.config.js
+++ b/tests/fixtures/base-1dep-optional/webpack.config.js
@@ -1,0 +1,19 @@
+var HardSourceWebpackPlugin = require('../../..');
+
+module.exports = {
+  context: __dirname,
+  entry: './index.js',
+  output: {
+    path: __dirname + '/tmp',
+    filename: 'main.js',
+  },
+  recordsPath: __dirname + '/tmp/cache/records.json',
+  plugins: [
+    new HardSourceWebpackPlugin({
+      cacheDirectory: 'cache',
+      environmentPaths: {
+        root: __dirname + '/../../..',
+      },
+    }),
+  ],
+};

--- a/tests/fixtures/base-context-optional/b/10.js
+++ b/tests/fixtures/base-context-optional/b/10.js
@@ -1,0 +1,1 @@
+module.exports = 10;

--- a/tests/fixtures/base-context-optional/b/6.js
+++ b/tests/fixtures/base-context-optional/b/6.js
@@ -1,0 +1,1 @@
+module.exports = 6;

--- a/tests/fixtures/base-context-optional/b/7.js
+++ b/tests/fixtures/base-context-optional/b/7.js
@@ -1,0 +1,1 @@
+module.exports = 7;

--- a/tests/fixtures/base-context-optional/b/8.js
+++ b/tests/fixtures/base-context-optional/b/8.js
@@ -1,0 +1,1 @@
+module.exports = 8;

--- a/tests/fixtures/base-context-optional/b/9.js
+++ b/tests/fixtures/base-context-optional/b/9.js
@@ -1,0 +1,1 @@
+module.exports = 9;

--- a/tests/fixtures/base-context-optional/b/index.js
+++ b/tests/fixtures/base-context-optional/b/index.js
@@ -1,0 +1,6 @@
+module.exports =
+  require('./6') +
+  require('./7') +
+  require('./8') +
+  require('./9') +
+  require('./10');

--- a/tests/fixtures/base-context-optional/index.js
+++ b/tests/fixtures/base-context-optional/index.js
@@ -1,0 +1,8 @@
+var a = 0;
+try {
+  var context = require.context('./a');
+  context.keys().forEach(function(m) {a += context(m);});
+}
+catch (_) {}
+
+console.log(a + require('./b'));

--- a/tests/fixtures/base-context-optional/webpack.config.js
+++ b/tests/fixtures/base-context-optional/webpack.config.js
@@ -1,0 +1,16 @@
+var HardSourceWebpackPlugin = require('../../..');
+
+module.exports = {
+  context: __dirname,
+  entry: './index.js',
+  output: {
+    path: __dirname + '/tmp',
+    filename: 'main.js',
+  },
+  recordsPath: __dirname + '/tmp/cache/records.json',
+  plugins: [
+    new HardSourceWebpackPlugin({
+      cacheDirectory: 'cache',
+    }),
+  ],
+};

